tests/test_sendinblue_backend.py

```diff
@@ -0,0 +1,444 @@
+            mail.send_mail('Subject', 'Body', 'from@example.com', ['to@example.com'])
+
+        # Make sure fail_silently is respected
+        self.set_mock_response(status_code=400)
+        sent = mail.send_mail('Subject', 'Body', 'from@example.com', ['to@example.com'], fail_silently=True)
+        self.assertEqual(sent, 0)
+
+    def test_api_error_includes_details(self):
+        """AnymailAPIError should include ESP's error message"""
+        # JSON error response:
+        error_response = b"""{
+            "code": "invalid_parameter",
+            "message": "valid sender email required"
+        }"""
+        self.set_mock_response(status_code=400, raw=error_response)
+        with self.assertRaises(AnymailAPIError) as cm:
+            self.message.send()
+        err = cm.exception
+        self.assertIn("code", str(err))
+        self.assertIn("message", str(err))
+
+        # No content in the error response:
+        self.set_mock_response(status_code=502, raw=None)
+        with self.assertRaises(AnymailAPIError):
+            self.message.send()
+
+
+class SendinBlueBackendAnymailFeatureTests(SendinBlueBackendMockAPITestCase):
+    """Test backend support for Anymail added features"""
+
+    def test_metadata(self):
+        self.message.metadata = {'user_id': "12345", 'items': 6, 'float': 98.6, 'long': longtype(123)}
+
+        with self.assertRaises(AnymailUnsupportedFeature):
+            self.message.send()
+
+    def test_send_at(self):
+        utc_plus_6 = get_fixed_timezone(6 * 60)
+        utc_minus_8 = get_fixed_timezone(-8 * 60)
+
+        with override_current_timezone(utc_plus_6):
+            # Timezone-aware datetime converted to UTC:
+            self.message.send_at = datetime(2016, 3, 4, 5, 6, 7, tzinfo=utc_minus_8)
+
+            with self.assertRaises(AnymailUnsupportedFeature):
+                self.message.send()
+
+    def test_tag(self):
+        self.message.tags = ["receipt"]
+        self.message.send()
+        data = self.get_api_call_json()
+        self.assertCountEqual(data['headers']["X-Mailin-tag"], "receipt")
+
+    def test_multiple_tags(self):
+        self.message.tags = ["receipt", "repeat-user"]
+        with self.assertRaises(AnymailUnsupportedFeature):
+            self.message.send()
+
+    def test_tracking(self):
+        # Test one way...
+        self.message.track_clicks = False
+        self.message.track_opens = True
+
+        with self.assertRaises(AnymailUnsupportedFeature):
+            self.message.send()
+
+        # ...and the opposite way
+        self.message.track_clicks = True
+        self.message.track_opens = False
+        with self.assertRaises(AnymailUnsupportedFeature):
+            self.message.send()
+
+    def test_template_id(self):
+        # SendinBlue use incremental ID to identify templates
+        self.message.template_id = "12"
+        self.message.send()
+
+        # We can't verify if the Url has changed
```

Comment:
`self.assert_esp_called('/v3/smtp/templates/12/send')` will check the URL. (See RequestsBackendMockAPITestCase.) 

This would also be a good place to verify that all the API parameters are transformed as expected for the template API call. ("to" --> "emailTo", etc.)

Reply:
Done